com.unity.multiplayer.mlapi/Runtime/Timing/ClientNetworkTimeProvider.cs

```diff
@@ -0,0 +1,121 @@
+
+        /// <summary>
+        /// Gets or sets the ratio at which the <see cref="ClientNetworkTimeProvider"/> speeds up or slows down time.
+        /// </summary>
+        public float AdjustmentRatio { get; set; } = 0.01f;
+
+        /// <summary>
+        /// Gets or sets a threshold in seconds used to force a hard catchup of network time.
+        /// </summary>
+        public float HardResetThreshold { get; set; } = 200f / 1000f; // 200ms
+
+        /// <summary>
+        /// Gets the current time scale for server time.
+        /// </summary>
+        public float ServerTimeScale { get; private set; } = 1f;
+
+        /// <summary>
+        /// Gets the current time scale for local time.
+        /// </summary>
+        public float LocalTimeScale { get; private set; } = 1f;
+
+        /// <summary>
+        /// Gets or sets the amount of time in seconds the server should buffer incoming client messages.
+        /// This increases the difference between local and server time so that messages arrive earlier on the server.
+        /// A hy
+        /// </summary>
+        public float TargetClientBufferTime { get; set; }
+
+        /// <summary>
+        /// Gets or sets the amount of the time in seconds the client should buffer incoming messages from the server. This increases server time.
+        /// A higher value increases latency but makes the game look more smooth in bad networking conditions.
+        /// </summary>
+        public float TargetServerBufferTime { get; set; }
+
+        /// <summary>
+        /// Creates a new instance of the <see cref="ClientNetworkTimeProvider"/> class with default buffer values.
+        /// </summary>
+        /// <param name="networkStats">The <see cref="INetworkStats"/> to get network information from.</param>
+        /// <param name="tickRate">The tickrate.</param>
+        public ClientNetworkTimeProvider(INetworkStats networkStats, int tickRate)
+        {
+            m_NetworkStats = networkStats;
+
+            // default buffer values. Can be manually overriden.
+            TargetClientBufferTime = 1f / tickRate;
+            TargetServerBufferTime = 2f / tickRate;
+        }
+
+        /// <inheritdoc/>
+        public bool AdvanceTime(ref NetworkTime localTime, ref NetworkTime serverTime, double deltaTime)
+        {
+            // advance time
+            localTime += deltaTime * LocalTimeScale;
+            serverTime += deltaTime * ServerTimeScale;
+
+            // time scale adjustment based on whether we are behind / ahead of the server in terms of inputs
+            // This implementation uses RTT to calculate that without server input which is not ideal. In the future we might want to add a field to the protocol which allows the server to send the exact input buffers size to the client.
+            double lastReceivedSnapshotTime = m_NetworkStats.LastReceivedSnapshotTick.Time;
+            double targetServerTime = lastReceivedSnapshotTime - TargetClientBufferTime;
```

Comment:
I'm trying to reason about this...and to be honest I can't remember from the spec.  Is `lastReceivedSnapshotTime` the time the server said it sent the packet or the time we recorded when we rcvd it?  I'm assuming it's the latter based on how `targetLocalTime` is computed.  

Did you consider specifying this in terms of ticks not time?  I guess I picture this more as a "number of tick slots"

Reply:
`lastReceivedSnapshotTime` is the tick we received with the last snapshot message (to convert it to a time value we simply multiply by the tick interval).